oauth2/src/main/java/com/linecorp/armeria/client/auth/oauth2/AbstractOAuth2AuthorizationGrant.java

```diff
@@ -0,0 +1,268 @@
+    @Nullable
+    private final Consumer<AccessTokenCapsule> tokenConsumer;
+
+    AbstractOAuth2AuthorizationGrant(RefreshAccessTokenRequest refreshRequest, Duration refreshBefore,
+                                     @Nullable Supplier<AccessTokenCapsule> tokenSupplier,
+                                     @Nullable Consumer<AccessTokenCapsule> tokenConsumer) {
+        tokenRef = new AtomicReference<>();
+        serialExecutor = Executors.newSingleThreadExecutor();
+        this.refreshRequest = requireNonNull(refreshRequest, "refreshRequest");
+        this.refreshBefore = requireNonNull(refreshBefore, "refreshBefore");
+        this.tokenSupplier = tokenSupplier;
+        this.tokenConsumer = tokenConsumer;
+    }
+
+    /**
+     * Obtains a new access token from the token end-point asynchronously.
+     * @return A {@link CompletableFuture} carrying the requested {@link AccessTokenCapsule} or an exception,
+     *         if the request failed.
+     * @throws TokenRequestException when the endpoint returns {code HTTP 400 (Bad Request)} status and the
+     *                               response payload contains the details of the error.
+     * @throws InvalidClientException when the endpoint returns {@code HTTP 401 (Unauthorized)} status, which
+     *                                typically indicates that client authentication failed (e.g.: unknown
+     *                                client, no client authentication included, or unsupported authentication
+     *                                method).
+     * @throws UnsupportedMediaTypeException if the media type of the response does not match the expected
+     *                                       (JSON).
+     */
+    protected abstract CompletableFuture<AccessTokenCapsule> obtainAccessTokenAsync(
+            @Nullable AccessTokenCapsule token);
+
+    /**
+     * Obtains a new access token from the token end-point.
+     * Optionally stores access token to registered {@link Consumer} for longer term storage.
+     * @return an {@link AccessTokenCapsule} that contains requested access token.
+     * @throws TokenRequestException when the endpoint returns {code HTTP 400 (Bad Request)} status and the
+     *                               response payload contains the details of the error.
+     * @throws InvalidClientException when the endpoint returns {@code HTTP 401 (Unauthorized)} status, which
+     *                                typically indicates that client authentication failed (e.g.: unknown
+     *                                client, no client authentication included, or unsupported authentication
+     *                                method).
+     * @throws UnsupportedMediaTypeException if the media type of the response does not match the expected
+     *                                       (JSON).
+     */
+    private AccessTokenCapsule obtainAccessToken() {
+        final AccessTokenCapsule token = obtainAccessTokenAsync(null).join();
```

Comment:
Ditto, Couldn't we return `CompletableFuture<AccessTokenCapsule>`?

Reply:
See my previous comment on the subject...
Yes, there is a special reason for synchronizing the request execution here.
In order to deal with concurrency issues, all token obtain and refresh requests have to be executed serially by a dedicated serialExecutor, which was introduced based on the previous @trustin comment. So the obtaining or refreshing of the access token is indeed happens asynchronously by the serialExecutor, but anything that happens during the refresh or obtain have to be serialized.

Reply:
Perhaps we could ensure serial execution without using a dedicated executor? For example, we could write some arbiter class that keeps the `CompletableFuture` of the previous task and add a callback to it to run the next task?

Reply:
I'm sure we can. We just need to find the right way. Let me think about your suggestion. We would still need another thread (or executor) to execute Obtain/Refresh, right? Your original comment was to avoid blocking the event loop.

Reply:
> We would still need another thread (or executor) to execute Obtain/Refresh, right?

If what 'obtain/refresh' does is simply doing some HTTP calls and waiting for them, we do not need another thread or executor. That's the whole point of going asynchronous. 😉 Do you have any specific task that will block for a long time during obtain/refresh? I'd be happy to help you make them fully asynchronous.

Reply:
> If what 'obtain/refresh' does is simply doing some HTTP calls and waiting for them, we do not need another thread or executor. That's the whole point of going asynchronous. 😉 Do you have any specific task that will block for a long time during obtain/refresh? I'd be happy to help you make them fully asynchronous.

Yes, I do. I have pointed at such example in the other thread. Some grants may involve continuous polling of the OAuth 2 token endpoint. For example [Device Authorization Flow](https://tools.ietf.org/html/rfc8628#section-3.5) which I have also implemented but did not include into this PR. This Flow assumes that the token obtained through a series or requests to which the OAuth 2 token endpoint responds with `authorization_pending` specifying some backoff or instructing to `slow_down`, and finally with the `access token` itself. This grant type is common to equipment activation (including consumer electronics), where the device starts the token `Obtain` cycle, while the authorization takes place independently by the user given the previously issued `User Code`. Many of us dealt with such grant type activating some apps on a smart TV, for instance.

```
      +----------+                                +----------------+
      |          |>---(A)-- Client Identifier --->|                |
      |          |                                |                |
      |          |<---(B)-- Device Code,      ---<|                |
      |          |          User Code,            |                |
      |  Device  |          & Verification URI    |                |
      |  Client  |                                |                |
      |          |  [polling]                     |                |
      |          |>---(E)-- Device Code       --->|                |
      |          |          & Client Identifier   |                |
      |          |                                |  Authorization |
      |          |<---(F)-- Access Token      ---<|     Server     |
      +----------+   (& Optional Refresh Token)   |                |
            v                                     |                |
            :                                     |                |
           (C) User Code & Verification URI       |                |
            :                                     |                |
            v                                     |                |
      +----------+                                |                |
      | End User |                                |                |
      |    at    |<---(D)-- End user reviews  --->|                |
      |  Browser |          authorization request |                |
      +----------+                                +----------------+

```

Reply:
Yes, such periodic check with backoff can be done completely asynchronously. That's how we do in `RetryingClient`. Of course it's more complicated than synchronous flow, but we'd like to make sure all this can be done in our event loop thread, so that a user doesn't have to manage another thread pool. How about focusing on making what's in this pull request asynchronous first? We could then discuss/cowork in more detail about how to make the second part asynchronous.

Reply:
Made a refactoring for this. Please see my other comment with a summary.